dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java

```diff
@@ -192,20 +193,23 @@ public void run() {
                 sendAlert(this.task.getTaskAlertInfo(), this.task.getExitStatus().getCode());
             }
 
-            taskExecutionContext.setCurrentExecutionStatus(ExecutionStatus.of(this.task.getExitStatus().getCode()));
-            taskExecutionContext.setEndTime(DateUtils.getCurrentDate());
-            taskExecutionContext.setProcessId(this.task.getProcessId());
-            taskExecutionContext.setAppIds(this.task.getAppIds());
             taskExecutionContext.setVarPool(JSONUtils.toJsonString(this.task.getParameters().getVarPool()));
-            logger.info("task instance id : {},task final status : {}", taskExecutionContext.getTaskInstanceId(), this.task.getExitStatus());
+
+            logger.info("task instance id : {}, task final status : {}", taskExecutionContext.getTaskInstanceId(), this.task.getExitStatus());
         } catch (Throwable e) {
-            logger.error("task scheduler failure", e);
+            logger.error("task instance id : {}, scheduler failure", taskExecutionContext.getTaskInstanceId(), e);
+            errorFlag = true;
             kill();
-            taskExecutionContext.setCurrentExecutionStatus(ExecutionStatus.FAILURE);
+        } finally {
+            if (errorFlag) {
+                taskExecutionContext.setCurrentExecutionStatus(ExecutionStatus.FAILURE);
+            } else {
+                taskExecutionContext.setCurrentExecutionStatus(ExecutionStatus.of(this.task.getExitStatus().getCode()));
+            }
             taskExecutionContext.setEndTime(DateUtils.getCurrentDate());
             taskExecutionContext.setProcessId(this.task.getProcessId());
```

Comment:
If in finally, the task maybe null, so it need to add NPE check.

Reply:
> If in finally, the task maybe null, so it need to add NPE check.

yes, I check again. you are right. if add NPE check, I think setting in try block and catch block is better. it readability is better. WDYT?

Reply:
How do I determine whether a task is null through executionStatus？There may be the same value` ExecutionStatus.FAILURE` in try block and catch block and task may also be null.  I think if task throw NPE, it need to add NPE check in both catch block and finally block . so I think add check in finally is better.

Reply:
You can define a ExecutionStatus and set the value at catch, then you don't need to parse the status from task.
```java
ExecutionStatus executionStatus;
try{
executionStatus = ExecutionStatus.of(this.task.getExitStatus().getCode());
}catch(Exception) {
executionStatus = ExecutionStatus.FAILURE;
}finally{
taskExecutionContext.setCurrentExecutionStatus(executionStatus);
}
```

Reply:
the task is also used in getProcessId() and getAppIds() , so it need to add NPE check